x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -325,23 +323,31 @@ public Authentication maybeAddAnonymousRoles(@Nullable AnonymousUser anonymousUs
         }
     }
 
+    // Package private for tests
     /**
      * Returns {@code true} if the effective user belongs to a realm under a domain.
-     * See also {@link #getDomain()} and {@link #getSourceRealm()}.
      */
-    public boolean isAssignedToDomain() {
-        return getDomain() != null;
+    boolean isAssignedToDomain() {
+        if (isFailedRuns()) {
+            return false;
+        }
+        return getEffectiveSubject().getRealm().getDomain() != null;
     }
 
+    // Package private for tests
     /**
      * Returns the {@link RealmDomain} that the effective user belongs to.
      * A user belongs to a realm which in turn belongs to a domain.
      *
      * The same username can be authenticated by different realms (e.g. with different credential types),
      * but resources created across realms cannot be accessed unless the realms are also part of the same domain.
      */
-    public @Nullable RealmDomain getDomain() {
-        return getSourceRealm().getDomain();
+    @Nullable
+    RealmDomain getDomain() {
```

Comment:
I changed this method (also `isAssignedToDomain`) to package private because:
1. It is not really used in production code
2. I am not sure whether we want them. Since we removed `getRealm` method from Authentication, having a `getDomain` feels going backwards.

Reply:
Agreed that it doesn't make sense to expose either of the domain methods beyond tests.